### ISSUE TITLE
Behavior of --prefer-stubs

### ISSUE THREAD

**Issue by furgerf** [pylint-dev/pylint#9889]
### Bug description

```python
Hi, when using `--prefer-stubs yes` I get lots of errors around numpy - I reported some examples [with numpy](https://github.com/numpy/numpy/issues/27244#issuecomment-2296962134).

I'm not sure what the expected behavior is of `--prefer-stubs` - it seems that with the flag set, `pylint` _exclusively_ looks at stub files but ignores other types. That's not what I'd expect from "prefer stubs", hence I'm filing this as a bug.

Ideally, I'd be able to selectively `--prefer-stubs` of certain modules, though that doesn't seem to be possible at the moment.
```

### Command used

```shell
pylint --prefer-stubs yes foo.py
```

### Pylint output

```python
************* Module foo
foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
foo.py:4:17: E1136: Value 'npt.NDArray' is unsubscriptable (unsubscriptable-object)
foo.py:6:20: E1133: Non-iterable value np.split(np.arange(12), 3) is used in an iterating context (not-an-iterable)
foo.py:6:4: R1721: Unnecessary use of a comprehension, use list(np.split(np.arange(12), 3)) instead. (unnecessary-comprehension)
foo.py:8:20: E1133: Non-iterable value np.diff(np.arange(12)) is used in an iterating context (not-an-iterable)
foo.py:8:4: R1721: Unnecessary use of a comprehension, use list(np.diff(np.arange(12))) instead. (unnecessary-comprehension)
foo.py:10:4: E1136: Value 'np.arange(12)' is unsubscriptable (unsubscriptable-object)
```

### Expected behavior

I expect to only see the errors around setting up the examples:

```
************* Module foo
foo.py:1:0: C0104: Disallowed name "foo" (disallowed-name)
foo.py:6:4: R1721: Unnecessary use of a comprehension, use list(np.split(np.arange(12), 3)) instead. (unnecessary-comprehension)
foo.py:8:4: R1721: Unnecessary use of a comprehension, use list(np.diff(np.arange(12))) instead. (unnecessary-comprehension)
```

### Pylint version

```shell
pylint 3.2.6
astroid 3.2.4
Python 3.11.2 (main, Feb  9 2023, 05:12:33) [GCC 10.2.1 20210110]
```

### OS / Environment

Docker (python:3.11-slim⁠ image)

**Comment by Pierre-Sassoulas** [comment]
I'm wondering if the path forward could be to "transform" stubs into astroid brains then act accordingly (at run time, I'm not advocating to create hard coded astroid brains from stubs that we'd then need to maintain)

**Comment by furgerf** [comment]
> Hi @furgerf, you're quite right that `--prefer-stubs` is naive and just exclusively looks at stubs. Part of the reason we immediately put the feature under a feature flag was that we felt we were responding to the requests of a few users and not truly understanding the full range of use cases. Unfortunately, there are no concrete plans to continue developing this feature. We need a champion to understand the use cases and propose some improvements.

Thanks for your reply - the way I understand it "either it works as-is for your use-case or you probably don't want to use it (at least for now)". Unfortunately, I don't think I can make a serious proposal, as mine is also just a single use-case and I lack quite a bit of background on what this would entail.
I don't know if you may be tracking something like this elsewhere already - if so, feel free to close this 🙂

**Comment by Pierre-Sassoulas** [comment]
There was a proposal in https://github.com/pylint-dev/pylint/issues/10017, that I'm going to regroup here:

> pylint should honor .pyi files even if there's no corresponding .py file
